### PR TITLE
[CINFRA-674][CINFRA-675]  Replicated operations refactoring

### PR DESCRIPTION
**Pull request by apetenchea**
### Summary

This pull request refactors DocumentLogEntries to use variants instead of enums and fixes several bugs that were discovered during the refactoring.

- Refactored DocumentLogEntries to use variants instead of enums
- Fixed several bugs that were discovered during the refactoring, including handling the IntermediateCommit operation
- Introduced a new ReplicatedOperation structure to indicate the command that the replicated state machine has to execute
- Fixed a potential problem caused by early compaction on the leader
- Separated the decision to replicate a transaction from replicateOperations and moved that logic into a separate method: needsReplication
- Added additional C++ tests as needed and refactored existing tests

### Gory details

1. We are refactoring `DocumentLogEntries` to use variants instead of enum. Due to this refactoring alone, several bugs were found and fixed, as pointed out by the compiler (for example, we forgot to handle the `IntermediateCommit` operation in several places). The `DocumentLogEntry` structure remains as the main mechanism for *transporting* log entries into the stream, while the newly added `ReplicatedOperation` *indicates the command* that the replicated state machine has to execute. Currently, the log entry is only a wrapper on top of the operation, but I believe this separation could prove to be beneficial in the future. This change is responsible for most lines modified in this PR, and brought with itself a necessary refactoring in the unit tests (both C++ and JS). Additional C++ tests were added as needed. This point address [CINFRA-674](https://arangodb.atlassian.net/browse/CINFRA-674).
2. A potential problem caused by early compaction was fixed on the leader. When dealing with `Commit` and `CreateShard` operations, the leader has to replicate them first, wait for the entries to be committed (in the log sense), commit the operation in RocksDB (or create the shard) and *only then* call release on the stream. This is now achieved by introducing a `release` method on the leader. There are currently two versions, one for handling user transactions that carry an official `TransactionID` and cannot simply be split in half by compaction, and one taking just a log index, which is enough for identifying an internal transaction (such as `DropShard`). This could be improved in the future, such that we no longer have to handle two different cases, perhaps we could generate transaction IDs for internal operations as well. The first plan was to use a `ScopeGuard` here, but I found that a bit tricky due to potential exceptions being thrown in `doCommit()`. Nevertheless, the focus here was to get a working version, while making sure the current refactoring allows for such improvements in the future. See [CINFRA-675](https://arangodb.atlassian.net/browse/CINFRA-675) for reference.
3. We separated the decision to replicate a transaction from `replicateOperations` and moved that logic into a separate method: `needsReplication`. This makes it more visible where are the places in which we decide what to replicate or not. However, as single commit/abort operations can still be encountered on the follower, we've made sure these are safely ignored. See the comments on [CINFRA-695](https://arangodb.atlassian.net/browse/CINFRA-694).
4. Some additional bug-fixes and minor TODO's were covered in the rest of this PR.

- [x] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [x] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [x] C++ **Unit tests**
  - [x] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.10: *(Please link PR)*
  - [ ] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*

[CINFRA-674]: https://arangodb.atlassian.net/browse/CINFRA-674?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CINFRA-675]: https://arangodb.atlassian.net/browse/CINFRA-675?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: arangod/Replication2/StateMachines/Document/DocumentStateTransactionHandler.cpp
Comment (mchacki) [inline arangod/Replication2/StateMachines/Document/DocumentStateTransactionHandler.cpp:120]: Same as before, we probably want to consider using override visitor here.
With this many cases it gets a bit hard to visually see that all `else if constexpr` are correct, and we not only have a single if at one place.

File: arangod/Replication2/StateMachines/Document/DocumentStateTransactionHandler.cpp
Comment (mchacki) [inline arangod/Replication2/StateMachines/Document/DocumentStateTransactionHandler.cpp:187]: Please wrap this 2 lines into ifdef MAINTAINER_MODE

The transactions are only used within the Assert, so we can help the compiler that this is actually only required in Maintainer mode, and not required in release. (We probably end up with "unused" compiler warning here anyhow.

File: arangod/Replication2/StateMachines/Document/ReplicatedOperation.cpp
Comment (mchacki) [inline arangod/Replication2/StateMachines/Document/ReplicatedOperation.cpp:5]:
```suggestion
/// Copyright 2023-2023 ArangoDB GmbH, Cologne, Germany
```

File: arangod/Replication2/StateMachines/Document/ReplicatedOperation.h
Comment (mchacki) [inline arangod/Replication2/StateMachines/Document/ReplicatedOperation.h:5]:
```suggestion
/// Copyright 2023-2023 ArangoDB GmbH, Cologne, Germany
```

File: arangod/Replication2/StateMachines/Document/ReplicatedOperation.h
Comment (mchacki) [inline arangod/Replication2/StateMachines/Document/ReplicatedOperation.h:40]: This comment holds for all of the following file.

Do we need these elements exactly where they are serialized/deserialzed as well?
Or are they passed around, and only serialized/deserialized in fewer files?

If the second is true, we should have a follow up task to split "Inspectors" from "Structs" into two separate header files.
The inspector has high impact on compile time, that we can avoid for files just using the structs, but have nothing to do with their outside representation

File: arangod/RocksDBEngine/ReplicatedRocksDBTransactionCollection.cpp
Comment (mchacki) [inline arangod/RocksDBEngine/ReplicatedRocksDBTransactionCollection.cpp:224]: Do we have a ticket for this?

**Review by mchacki — CHANGES_REQUESTED** [review]
In general a very good PR.

My request change is centered around the TRI_ASSERT with side-effect, I think it is important to get this sorted out before merge. (Otherwise TestCode and ReleaseCode have different behavior)

All other comments can be considered nice to have, and I would not block the PR for any of them.

File: arangod/Replication2/StateMachines/Document/ActiveTransactionsQueue.cpp
Comment (apetenchea) [reply arangod/Replication2/StateMachines/Document/ActiveTransactionsQueue.cpp:85]: Sure, doesn't hurt to have an ADB_PROD_ASSERT

File: arangod/Replication2/StateMachines/Document/DocumentFollowerState.cpp
Comment (apetenchea) [reply arangod/Replication2/StateMachines/Document/DocumentFollowerState.cpp:106]: I think the overloaded version looks nicer indeed, but my goal was to make sure that I handle all cases, which now I can using the `static_assert(always_false_v<T>, "non-exhaustive visitor!");` at line 189. Unfortunately I don't think that such an assertion is possible with the overloaded approach.

File: arangod/Replication2/StateMachines/Document/DocumentFollowerState.h
Comment (apetenchea) [reply arangod/Replication2/StateMachines/Document/DocumentFollowerState.h:85]: You are right, it's only being used inside the `_guardedData` lock anyway.

File: arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp
Comment (apetenchea) [reply arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp:198]: I used `auto&&` because afaik in that case the compiler can choose whatever fits best.

File: arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp
Comment (apetenchea) [reply arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp:415]: Proper deletion of snapshots will be handled with [CINFRA-676](https://arangodb.atlassian.net/browse/CINFRA-676)

File: arangod/Replication2/StateMachines/Document/DocumentStateShardHandler.h
Comment (apetenchea) [reply arangod/Replication2/StateMachines/Document/DocumentStateShardHandler.h:73]: I wanted to use a `shared_lock` for reading data, which I didn't manage to implement with a `GuardedData`. I don't think it can differentiate between *shared* and *unique* locking.

File: arangod/Replication2/StateMachines/Document/DocumentStateTransaction.cpp
Comment (apetenchea) [reply arangod/Replication2/StateMachines/Document/DocumentStateTransaction.cpp:61]: I created one: [CINFRA-697](https://arangodb.atlassian.net/browse/CINFRA-697)

File: arangod/Replication2/StateMachines/Document/DocumentStateTransactionHandler.cpp
Comment (apetenchea) [reply arangod/Replication2/StateMachines/Document/DocumentStateTransactionHandler.cpp:120]: My problem is that I'm trying to have these `static_assert(always_false_v<T>, "Unhandled operation type. This should never happen.");` help me make sure that I handle all cases at compile time. In the end, the overload solution would work fine for me, I don't have a strong preference on how to achieve this visitor thing. We could bring that as a topic of debate to one of our team meetings and decide upon a common approach.

File: arangod/Replication2/StateMachines/Document/DocumentStateTransactionHandler.cpp
Comment (apetenchea) [reply arangod/Replication2/StateMachines/Document/DocumentStateTransactionHandler.cpp:187]: Good catch!

File: arangod/RocksDBEngine/ReplicatedRocksDBTransactionCollection.cpp
Comment (apetenchea) [reply arangod/RocksDBEngine/ReplicatedRocksDBTransactionCollection.cpp:224]: Created one: [CINFRA-698](https://arangodb.atlassian.net/browse/CINFRA-698)

File: arangod/Replication2/StateMachines/Document/ReplicatedOperation.h
Comment (apetenchea) [reply arangod/Replication2/StateMachines/Document/ReplicatedOperation.h:39]: That's a good observation, I have split them into separate files!

File: arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp
Comment (apetenchea) [reply arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp:236]: I expanded the current comment

File: arangod/Replication2/StateMachines/Document/DocumentStateShardHandler.h
Comment (apetenchea) [reply arangod/Replication2/StateMachines/Document/DocumentStateShardHandler.h:73]: I created a ticket for that: [CINFRA-701](https://arangodb.atlassian.net/browse/CINFRA-701)

File: arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp
Comment (apetenchea) [reply arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp:107]: The problem is a bit more complicated for `recoverEntries`, because there are several places where we return early from the `std::visit` call. There is also some common code at the end, the `transactionHandler->applyEntry` call, which might happen or not, depending on the logic contained in the if statements above.

**Comment by apetenchea** [comment]
I created [CINFRA-702](https://arangodb.atlassian.net/browse/CINFRA-702) for ideas of future improvements.

File: arangod/Replication2/StateMachines/Document/DocumentFollowerState.cpp
Comment (mpoeter) [inline arangod/Replication2/StateMachines/Document/DocumentFollowerState.cpp:104]:
```suggestion
            auto [index, doc] = *entry;

            auto currentReleaseIndex = std::visit(
                [&data, index](
```

File: arangod/Replication2/StateMachines/Document/DocumentFollowerState.cpp
Comment (mpoeter) [inline arangod/Replication2/StateMachines/Document/DocumentFollowerState.cpp:117]: Looks like `releaseIndex` is only set here, so we don't really need it and can return `currentReleaseIndex` instead.

File: arangod/Replication2/StateMachines/Document/DocumentFollowerState.cpp
Comment (mpoeter) [inline arangod/Replication2/StateMachines/Document/DocumentFollowerState.cpp:353]: Theoretically we can release everything in this transaction up to this point and update the start LogIndex of this transaction to the current LogIndex. However, I would not do this now but maybe put it down so we can revisit this later, perhaps in the refactoring where we try to unify code between the follower state and the leader recovery.

File: arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp
Comment (mpoeter) [inline arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp:168]: Do we really need to block here until the entry is replicated?

**Review by mchacki — APPROVED** [review]
LGTM

File: arangod/Replication2/StateMachines/Document/DocumentStateTransactionHandler.cpp
Comment (mpoeter) [inline arangod/Replication2/StateMachines/Document/DocumentStateTransactionHandler.cpp:36]: NIT: any particular reason why you moved this function into the DocumentStateTransactionHandler? I usually keep function that are not really related to a class (low cohesion, ie., do not rely on its members/functions) outside as separate functions, especially when they are not used outside of this module. This also helps keep the class' interface clean since a member function also has to be declared in the header.

File: arangod/Replication2/StateMachines/Document/DocumentStateTransactionHandler.cpp
Comment (mpoeter) [inline arangod/Replication2/StateMachines/Document/DocumentStateTransactionHandler.cpp:63]: NIT: same as above

File: arangod/Replication2/StateMachines/Document/DocumentStateTransactionHandler.cpp
Comment (mpoeter) [inline arangod/Replication2/StateMachines/Document/DocumentStateTransactionHandler.cpp:110]: Should we make this an `ADB_PROD_ASSERT`?

File: arangod/Replication2/StateMachines/Document/DocumentStateTransactionHandler.cpp
Comment (mpoeter) [inline arangod/Replication2/StateMachines/Document/DocumentStateTransactionHandler.cpp:160]: I would prefer something like this:
```suggestion
  auto res = std::invoke(
    overload{
      [&](ReplicatedOperation::Commit const&) { return trx->commit(); },
      [&](ReplicatedOperation::Abort const&) { return trx->abort); }
    }, op);
```
I think this is a bit easier to read and it would also cause a compile error if we ever add a new type to `FinishesUserTransaction`.

File: arangod/Transaction/Methods.cpp
Comment (mpoeter) [inline arangod/Transaction/Methods.cpp:2657]: Why should this finish immediately? Shouldn't this be async?

File: arangod/Transaction/Methods.cpp
Comment (mpoeter) [inline arangod/Transaction/Methods.cpp:3153]: Why should this finish immediately? Shouldn't this be async?

File: arangod/Replication2/StateMachines/Document/DocumentFollowerState.cpp
Comment (maierlars) [inline arangod/Replication2/StateMachines/Document/DocumentFollowerState.cpp:101]: copy of document log entry might be expensive if we write huge batches.

**Review by mpoeter — APPROVED** [review]

File: arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp
Comment (maierlars) [reply arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp:168]: I don't think so.

File: arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp
Comment (maierlars) [inline arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp:201]: This comment was confusing. Is it about not replicating the commit or abort for an empty transaction?

File: arangod/Replication2/StateMachines/Document/DocumentFollowerState.cpp
Comment (maierlars) [reply arangod/Replication2/StateMachines/Document/DocumentFollowerState.cpp:101]: forget about that, the log entry contains a `ShardSlice` which is cheap to copy.

File: arangod/Replication2/StateMachines/Document/DocumentFollowerState.cpp
Comment (apetenchea) [reply arangod/Replication2/StateMachines/Document/DocumentFollowerState.cpp:117]: I use `releaseIndex` in order to preserve only the last `currentReleaseIndex` that's calculated in the while loop, I can't just return `currentReleaseIndex` directly.

File: arangod/Replication2/StateMachines/Document/DocumentFollowerState.cpp
Comment (apetenchea) [reply arangod/Replication2/StateMachines/Document/DocumentFollowerState.cpp:353]: Good, noted that

File: arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp
Comment (apetenchea) [reply arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp:168]: We are not actually blocking, because we don't wait for commit. I will add an assertion to make sure that the future is always ready after the `replicateOperation` call.

Comment (maierlars) [reply arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp:168]: ah because `waitForCommit` is not set in the options 😉

File: arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp
Comment (apetenchea) [reply arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp]: Updated the comment

File: arangod/Transaction/Methods.cpp
Comment (apetenchea) [reply arangod/Transaction/Methods.cpp:2657]: It finishes immediately because we are not waiting for the operation to be committed in the log. I added an assertion and comment on that.

File: arangod/Replication2/StateMachines/Document/DocumentStateTransactionHandler.cpp
Comment (apetenchea) [reply arangod/Replication2/StateMachines/Document/DocumentStateTransactionHandler.cpp:160]: I updated the method.

File: arangod/Transaction/Methods.cpp
Comment (apetenchea) [reply arangod/Transaction/Methods.cpp:3153]: Same as above

File: arangod/Replication2/StateMachines/Document/DocumentStateTransactionHandler.cpp
Comment (apetenchea) [reply arangod/Replication2/StateMachines/Document/DocumentStateTransactionHandler.cpp:36]: While working with concepts and visitors, I had some linker issues which I wanted to get out of the way quickly, so I moved these functions around. That's no longer the case now, hence I moved them back to an anonymous namespace in the cpp file. I agree with your points.

File: arangod/Replication2/StateMachines/Document/DocumentFollowerState.cpp
Comment (mpoeter) [reply arangod/Replication2/StateMachines/Document/DocumentFollowerState.cpp:117]: Ah, good point. 🤦‍♂️

**Pull request merged**
